### PR TITLE
feat(react-components): basic echarts

### PR DESCRIPTION
**Pull request by corteggiano**
## Overview
- base react eCharts (wrapper courtesy of [https://dev.to/manufac/using-apache-echarts-with-react-and-typescript-353k](https://dev.to/manufac/using-apache-echarts-with-react-and-typescript-353k) )
- default eCharts options
- utils for echarts - `dataStreamToEChartseries`
- paginated legends, tooltips

new charts only used in the react-components storybook

## What is NOT supported yet?
- live data mode
- thresholds
- displaying selected aggregation type

Screenshot 2023-06-13 at 12 38 47
Screenshot 2023-06-13 at 13 22 25
Screenshot 2023-06-13 at 13 22 19
image

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

**Review by mnischay — COMMENTED** [review]
I am a bit hesitant to use it for the following reasons
1. its been 2 years since echarts-for-react was last published 
2. not sure if it exposes all the options/utilities echarts has  because we need to add custom tooltips example : getInstance().getHeight() 
3. do we really need another layer of abstraction ?

File: packages/react-components/src/utils/eChartsUtils.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import { DataStream } from '@iot-app-kit/charts-core';
+import { DataPoint, HistoricalViewport, Viewport } from '@iot-app-kit/core';
+
+export const dataStreamsToEChartSeries = (dataStreams: DataStream[], chartType: string) => {
+  return dataStreams.map((stream) => ({
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/utils/eChartsUtils.ts]: nice catch! filtering out alarm streams for now until we add the option to visualize them in eCharts :)

File: packages/react-components/src/index.ts [hunk 2]
```diff
@@ -14,6 +14,8 @@ export { KPI } from './components/kpi/kpi';
 export { StatusTimeline } from './components/status-timeline';
 export { Status } from './components/status/status';
 
+export { BaseECharts } from './components/base-e-chart';
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/index.ts]: woops left over from my testing 😄 removed!

File: packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  useEffect(() => {
+    let chart: ECharts | undefined;
+    if (chartRef.current !== null) {
+      chart = init(chartRef.current, theme, { width: 270, height: 250 });
+    }
+
+    function resizeChart() {
+      chart?.resize();
+    }
+    window.addEventListener('resize', resizeChart);
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx]: ended up removing this in favor of using the existing dashboard sizing logic!
now just passing through the calculated height and width of the parent container to the echarts component 😎

File: packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import { DEFAULT_SIZE } from './eChartsConstants';
+
+export interface EChartsWrapperProps {
+  option: EChartsOption;
+  style?: CSSProperties;
+  settings?: SetOptionOpts;
+  loading?: boolean;
+  theme?: string;
+  className?: string;
+  size?: any;
```
Comment (mnischay) [inline packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx:15]: probably should not be using "any" here...

File: packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+export function EChartsWrapper({ option, style, settings, loading, theme, size }: EChartsWrapperProps): JSX.Element {
```
Comment (mnischay) [inline packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx:18]: super nit: i think we follow arrow function notation ?

File: packages/react-components/src/utils/eChartsUtils.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { DataStream } from '@iot-app-kit/charts-core';
+import { DataPoint, HistoricalViewport, Viewport } from '@iot-app-kit/core';
+import type { SeriesOption } from 'echarts';
+import { StreamType } from '../common/constants';
+
+export const dataStreamsToEChartSeries = (dataStreams: DataStream[], chartType: string): SeriesOption[] => {
+  return dataStreams
+    .filter(({ streamType }) => streamType !== StreamType.ALARM)
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/utils/eChartsUtils.ts:8]: yep! in synchro-charts we always filter out alarm streams in line, bar, and scatter charts
it'll be configurable for other charts but in this case it wont be an option

File: packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import { init, getInstanceByDom } from 'echarts';
+
+import type { CSSProperties } from 'react';
+import type { ECharts, EChartsOption, SetOptionOpts } from 'echarts';
+
+export interface EChartsWrapperProps {
+  option: EChartsOption;
+  style?: CSSProperties;
+  settings?: SetOptionOpts;
+  loading?: boolean;
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx:12]: this component exposes the [ECharts+EChartsInstance API](https://echarts.apache.org/en/api.html#echartsInstance), which has all these props and some more

just getting ahead of adding some of the ones we'll likely want to use :)

File: packages/dashboard/src/hooks/useAssetDescriptionQueries.ts [hunk 8]
```diff
@@ -17,16 +17,15 @@ export const useAssetDescriptionMapQuery = (siteWiseAssetQuery: SiteWiseAssetQue
   const { iotSiteWiseClient } = useClients();
 
   const fetchSiteWiseAssetQueryDescription = async () => {
-    if (!iotSiteWiseClient || !siteWiseAssetQuery) return;
-
+    if (!iotSiteWiseClient || !siteWiseAssetQuery) return [];
```
Comment (corteggiano) [inline packages/dashboard/src/hooks/useAssetDescriptionQueries.ts:20]: returning undefined erroring and clogging up the console 
Screenshot 2023-06-16 at 08 27 48

File: packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  theme?: string;
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/components/base-e-chart/eChartsWrapper.tsx]: will do once the appkit config hook is merged in!

File: packages/react-components/src/components/chart/eChartsWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  theme?: string;
+  className?: string;
+  size?: { width?: number; height?: number };
+}
+
+export const EChartsWrapper = ({ option, settings, loading, theme, size = {} }: EChartsWrapperProps): JSX.Element => {
+  const chartRef = useRef<HTMLDivElement>(null);
+
+  // Initialize chart + listeners
+  useEffect(() => {
```
Comment (jmbuss) [inline packages/react-components/src/components/chart/eChartsWrapper.tsx:19]: I think we should turn these into hooks. I don't think we'd need the getInstanceByDom if we turn this into something like useECharts and it returns the chart?

Comment (jmbuss) [reply packages/react-components/src/components/chart/eChartsWrapper.tsx:19]: could probably also stuff all the resizing sideeffects in there too

File: packages/react-components/src/components/chart/eChartsWrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    if (chartRef.current !== null) {
+      const chart = getInstanceByDom(chartRef.current);
+      chart?.setOption(option, settings);
+    }
+  }, [chartRef, option, settings, theme]);
+
+  useEffect(() => {
+    if (chartRef.current !== null) {
+      const chart = getInstanceByDom(chartRef.current);
+      loading === true ? chart?.showLoading() : chart?.hideLoading();
```
Comment (jmbuss) [inline packages/react-components/src/components/chart/eChartsWrapper.tsx:41]: maybe we want a custom element, but I think echarts has some loading options - https://echarts.apache.org/en/api.html#echartsInstance.showLoading - have not messed around with it though

File: packages/react-components/src/components/chart/index.tsx [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+      ...DEFAULT_ECHART_OPTIONS,
+      xAxis: { ...DEFAULT_X_AXIS, show: axis?.showX ?? true },
+      yAxis: {
+        ...DEFAULT_Y_AXIS,
+        show: axis?.showY ?? true,
+        min: yMin,
+        max: yMax,
+      },
+      title: {
+        text: series.length === 0 ? `${chartType} chart - no data present` : '',
```
[the remark this answers is not in the document]
Comment (corteggiano) [reply packages/react-components/src/components/chart/index.tsx]: talked with Brian and we're most likely leaving this for contractors to handle!

**Review by jmbuss — APPROVED** [review]

File: packages/react-components/src/components/BaseECharts/BaseECharts.tsx [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    },
+    styles,
+  });
+
+  const series = useMemo(
+    () => dataStreamsToEChartSeries(dataStreams as DataStream[], chartType),
+    [dataStreams, chartType]
+  );
+
+  const option = useMemo(
```
Comment (jmbuss) [inline packages/react-components/src/components/BaseECharts/BaseECharts.tsx:75]: Maybe these should also be in the hook?

**Pull request merged**